File: python/lexscope/mcp.py
"""MCP stdio shim: local JSON-RPC 2.0 server exposing only documents_read and
records_delete. Newline-delimited UTF-8 JSON on stdio; batches rejected;
65536-byte message cap; JSON-RPC only on stdout; scrubbed diagnostics on
stderr. No credentials ever appear in tool schemas, results, or logs.
"""
from __future__ import annotations

import json
import sys

from .canon import parse_strict
from .ids import random_id
from .schemas import TOOL_NAME_RE
from .sdk import LexScopeError

PROTOCOL_VERSION = "2025-06-18"
MAX_MSG = 65536

TOOL_MAP = {
    "documents_read": "documents.read",
    "records_delete": "records.delete",
}

TOOL_DEFS = [
    {
        "name": "documents_read",
        "description": "Read one document in the configured workspace.",
        "inputSchema": {
            "type": "object",
            "additionalProperties": False,
            "required": ["workspace", "path"],
            "properties": {"workspace": {"type": "string"}, "path": {"type": "string"}},
        },
    },
    {
        "name": "records_delete",
        "description": "Delete one version-matched record in the configured workspace.",
        "inputSchema": {
            "type": "object",
            "additionalProperties": False,
            "required": ["workspace", "record_id", "expected_version"],
            "properties": {
                "workspace": {"type": "string"},
                "record_id": {"type": "string"},
                "expected_version": {"type": "integer", "minimum": 0},
            },
        },
    },
]


def _valid_id(i):
    if i is None:
        return True
    if isinstance(i, str):
        return 1 <= len(i) <= 64 and all(0x20 <= ord(c) <= 0x7E for c in i)
    if isinstance(i, int) and not isinstance(i, bool):
        return 0 <= i <= 9007199254740991
    return False


class McpServer:
    def __init__(self, client, handle, task_id, tools=None, idgen=None, out=None, diag=None):
        self.client = client
        self.handle = handle
        self.task_id = task_id
        self.tools = tools if tools is not None else ["documents.read", "records.delete"]
        self.idgen = idgen or random_id
        self.out = out or (lambda line: sys.stdout.write(line + "\n"))
        self.diag = diag or (lambda line: sys.stderr.write(line + "\n"))
        self.initialized = False

    def _send(self, msg):
        self.out(json.dumps(msg, separators=(",", ":")))

    def _reply(self, i, result):
        self._send({"jsonrpc": "2.0", "id": i, "result": result})

    def _err(self, i, code, message):
        self._send({"jsonrpc": "2.0", "id": i, "error": {"code": code, "message": message}})

    # one NDJSON line -> optional response lines. Never raises.
    def handle_line(self, line: str):
        if len(line.encode("utf-8")) > MAX_MSG:
            self._err(None, -32600, "message too large")
            return
        try:
            msg = parse_strict(line)
        except Exception:
            self._err(None, -32700, "parse error")
            return
        if isinstance(msg, list):
            self._err(None, -32600, "batches are not accepted")
            return
        if not isinstance(msg, dict):
            self._err(None, -32600, "invalid request")
            return
        if msg.get("jsonrpc") != "2.0" or not isinstance(msg.get("method"), str) or not _valid_id(msg.get("id")):
            self._err(None, -32600, "invalid request")
            return
        i = msg.get("id")
        method = msg["method"]
        params = msg.get("params")
        is_notif = "id" not in msg

        if method == "initialize":
            if i is None:
                return
            req = params.get("protocolVersion") if isinstance(params, dict) else None
            requested = req if isinstance(req, str) else None
            self.initialized = True
            result = {
                "protocolVersion": PROTOCOL_VERSION,
                "capabilities": {"tools": {}},
                "serverInfo": {"name": "lexscope", "version": "1.0.0"},
            }
            if requested is not None and requested != PROTOCOL_VERSION:
                result["requestedProtocolVersion"] = requested
            self._reply(i, result)
            return
        if method == "notifications/initialized":
            return  # notification: no response
        if method == "ping":
            if i is not None:
                self._reply(i, {})
            return
        if method == "tools/list":
            if not self.initialized:
                self._err(i, -32002, "not initialized")
                return
            # inventory is derived from the token's permitted tool names
            self._reply(i, {"tools": [d for d in TOOL_DEFS if TOOL_MAP[d["name"]] in self.tools]})
            return
        if method == "tools/call":
            if not self.initialized:
                self._err(i, -32002, "not initialized")
                return
            if not isinstance(params, dict):
                self._err(i, -32602, "invalid params")
                return
            name = params.get("name")
            args = params.get("arguments")
            if not isinstance(name, str) or name not in TOOL_MAP:
                self._err(i, -32602, "unknown tool")
                return
            if not isinstance(args, dict):
                self._err(i, -32602, "invalid arguments")
                return
            tool = TOOL_MAP[name]
            allowed = ["workspace", "path"] if tool == "documents.read" else ["workspace", "record_id", "expected_version"]
            for k in args:
                if k not in allowed:
                    self._err(i, -32602, f"unexpected argument {k}")
                    return
            call = {"v": 1, "call_id": self.idgen("lcl", 21), "task_id": self.task_id, "tool": tool, "args": args}
            try:
                res = self.client.call(self.handle, call)
                state = res.get("state") or res.get("outcome") or "UNKNOWN"
                text = json.dumps(res.get("result") if state == "SUCCEEDED" else {"outcome": state})
                self._reply(i, {"content": [{"type": "text", "text": text}], "isError": state != "SUCCEEDED"})
            except LexScopeError as e:
                self._reply(i, {"content": [{"type": "text", "text": json.dumps({"outcome": e.outcome, "code": e.code})}], "isError": True})
            except Exception:
                self.diag("tool call failed")
                self._reply(i, {"content": [{"type": "text", "text": json.dumps({"outcome": "UNKNOWN"})}], "isError": True})
            return
        if not is_notif:
            self._err(i, -32601, "method not found")

    # stdio entry: NDJSON in, NDJSON out. Diagnostics go to stderr, scrubbed.
    def run_stdio(self):
        for line in sys.stdin:
            line = line.rstrip("\n")
            try:
                self.handle_line(line)
            except Exception:
                self.diag("internal error")

File: src/mcp.ts
// MCP stdio shim: local JSON-RPC 2.0 server exposing only documents_read and
// records_delete. Newline-delimited UTF-8 JSON on stdio; batches rejected;
// 65536-byte message cap; JSON-RPC only on stdout; scrubbed diagnostics on
// stderr. No credentials ever appear in tool schemas, results, or logs.

import { createInterface } from "node:readline";
import { parseStrict } from "./core/strictjson.ts";
import { randomId } from "./core/ids.ts";
import type { LexScopeClient, TokenHandle } from "./sdk.ts";
import { LexScopeError } from "./sdk.ts";
import type { ToolCall, ToolName } from "./core/schemas.ts";

const PROTOCOL_VERSION = "2025-06-18";
const MAX_MSG = 65536;

const TOOL_MAP: Record<string, ToolName> = {
  documents_read: "documents.read",
  records_delete: "records.delete",
};

const TOOL_DEFS = [
  {
    name: "documents_read",
    description: "Read one document in the configured workspace.",
    inputSchema: {
      type: "object",
      additionalProperties: false,
      required: ["workspace", "path"],
      properties: {
        workspace: { type: "string" },
        path: { type: "string" },
      },
    },
  },
  {
    name: "records_delete",
    description: "Delete one version-matched record in the configured workspace.",
    inputSchema: {
      type: "object",
      additionalProperties: false,
      required: ["workspace", "record_id", "expected_version"],
      properties: {
        workspace: { type: "string" },
        record_id: { type: "string" },
        expected_version: { type: "integer", minimum: 0 },
      },
    },
  },
];

type JsonId = string | number | null;

function validId(id: unknown): id is JsonId {
  if (id === null) return true;
  if (typeof id === "string") return id.length >= 1 && id.length <= 64 && /^[\x20-\x7e]+$/.test(id);
  if (typeof id === "number") return Number.isSafeInteger(id) && id >= 0;
  return false;
}

export class McpServer {
  private client: LexScopeClient;
  private handle: TokenHandle;
  private taskId: string;
  private tools: ToolName[];
  private idgen: (p: string, l: number) => string;
  private initialized = false;
  private out: (line: string) => void;
  private diag: (line: string) => void;
  private seq = 0;

  constructor(opts: {
    client: LexScopeClient;
    handle: TokenHandle;
    taskId: string;
    tools?: ToolName[];
    idgen?: (p: string, l: number) => string;
    out?: (line: string) => void;
    diag?: (line: string) => void;
  }) {
    this.client = opts.client;
    this.handle = opts.handle;
    this.taskId = opts.taskId;
    this.tools = opts.tools ?? ["documents.read", "records.delete"];
    this.idgen = opts.idgen ?? ((p, l) => randomId(p, l));
    this.out = opts.out ?? ((l) => process.stdout.write(l + "\n"));
    this.diag = opts.diag ?? ((l) => process.stderr.write(l + "\n"));
  }

  private send(msg: unknown): void {
    this.out(JSON.stringify(msg));
  }
  private reply(id: JsonId, result: unknown): void {
    this.send({ jsonrpc: "2.0", id, result });
  }
  private err(id: JsonId, code: number, message: string): void {
    this.send({ jsonrpc: "2.0", id, error: { code, message } });
  }

  // one NDJSON line -> optional response lines. Never throws.
  async handleLine(line: string): Promise<void> {
    if (Buffer.byteLength(line, "utf8") > MAX_MSG) {
      this.err(null, -32600, "message too large");
      return;
    }
    let msg: unknown;
    try {
      msg = parseStrict(new TextEncoder().encode(line));
    } catch {
      this.err(null, -32700, "parse error");
      return;
    }
    if (Array.isArray(msg)) {
      this.err(null, -32600, "batches are not accepted");
      return;
    }
    if (typeof msg !== "object" || msg === null) {
      this.err(null, -32600, "invalid request");
      return;
    }
    const o = msg as Record<string, unknown>;
    if (o.jsonrpc !== "2.0" || typeof o.method !== "string" || !validId(o.id ?? null)) {
      this.err(null, -32600, "invalid request");
      return;
    }
    const id = (o.id ?? null) as JsonId;
    const method = o.method as string;
    const params = o.params;
    const isNotif = !("id" in o);

    switch (method) {
      case "initialize": {
        if (id === null) break;
        const req = typeof params === "object" && params !== null ? (params as Record<string, unknown>).protocolVersion : undefined;
        const requested = typeof req === "string" ? req : null;
        this.initialized = true;
        this.reply(id, {
          protocolVersion: PROTOCOL_VERSION,
          capabilities: { tools: {} },
          serverInfo: { name: "lexscope", version: "1.0.0" },
          ...(requested !== null && requested !== PROTOCOL_VERSION ? { requestedProtocolVersion: requested } : {}),
        });
        return;
      }
      case "notifications/initialized":
        return; // notification: no response
      case "ping":
        if (id !== null) this.reply(id, {});
        return;
      case "tools/list": {
        if (!this.initialized) { this.err(id, -32002, "not initialized"); return; }
        // inventory is derived from the token's permitted tool names
        this.reply(id, { tools: TOOL_DEFS.filter((d) => this.tools.includes(TOOL_MAP[d.name]!)) });
        return;
      }
      case "tools/call": {
        if (!this.initialized) { this.err(id, -32002, "not initialized"); return; }
        if (typeof params !== "object" || params === null) { this.err(id, -32602, "invalid params"); return; }
        const p = params as Record<string, unknown>;
        const name = p.name;
        const args = p.arguments;
        if (typeof name !== "string" || TOOL_MAP[name] === undefined) { this.err(id, -32602, "unknown tool"); return; }
        if (typeof args !== "object" || args === null || Array.isArray(args)) { this.err(id, -32602, "invalid arguments"); return; }
        const tool = TOOL_MAP[name];
        const allowed = tool === "documents.read" ? ["workspace", "path"] : ["workspace", "record_id", "expected_version"];
        for (const k of Object.keys(args)) {
          if (!allowed.includes(k)) { this.err(id, -32602, `unexpected argument ${k}`); return; }
        }
        const call = { v: 1, call_id: this.idgen("lcl", 21), task_id: this.taskId, tool, args } as ToolCall;
        this.seq++;
        try {
          const res = (await this.client.call(this.handle, call)) as { state?: string; outcome?: string; result?: unknown };
          const state = typeof res.state === "string" ? res.state : typeof res.outcome === "string" ? res.outcome : "UNKNOWN";
          const text = state === "SUCCEEDED" ? JSON.stringify(res.result ?? null) : JSON.stringify({ outcome: state });
          this.reply(id, { content: [{ type: "text", text }], isError: state !== "SUCCEEDED" });
        } catch (e) {
          if (e instanceof LexScopeError) {
            this.reply(id, { content: [{ type: "text", text: JSON.stringify({ outcome: e.outcome, code: e.code }) }], isError: true });
          } else {
            this.diag("tool call failed");
            this.reply(id, { content: [{ type: "text", text: JSON.stringify({ outcome: "UNKNOWN" }) }], isError: true });
          }
        }
        return;
      }
      default:
        if (!isNotif) this.err(id, -32601, "method not found");
    }
  }

  // stdio entry: NDJSON in, NDJSON out. Diagnostics go to stderr, scrubbed.
  runStdio(): void {
    const rl = createInterface({ input: process.stdin, terminal: false });
    rl.on("line", (line) => {
      this.handleLine(line).catch(() => this.diag("internal error"));
    });
  }
}
